表示を要素にしてクリック時にファイルのパスをコピーする

ShuntaToda opened the pull request:
初めまして、[戸田](https://x.com/shuntemskills)と申します。
https://zenn.dev/praha/articles/b2e225ae091ae3 の記事を拝見し、大変興味深く、便利なアプリだと感じました。

## 自己紹介
私はZennとDevelopersIOを運営しているクラスメソッド株式会社に所属しており、本年4月に新卒として入社いたしました。
なお、今回のご連絡は会社とは無関係の、個人的なものであることをご了承ください。

## 提案内容
まずZennの記事を読んで第一印象ですごく便利なアプリだと感じました。
ブログ執筆に以下の機能があると便利に感じたので実装しました。
1. ディレクトリ構造の表示をテキストから要素に変更
2. ディレクトリもしくはファイルをクリックした際に、そのパスをクリップボードにコピーする機能

これらの機能により、ユーザーの利便性が向上すると考えています。

## 今回のプルリクエストについて
実装した機能をGitHubのプルリクエストとして提出させていただきました。
まだ実務経験が浅く、綺麗なコードではないかもしれませんが、ご確認いただけますと幸いです。
もしご検討いただけるようでしたら、本番環境へのデプロイもご検討いただければ大変ありがたいです。

ご多忙中恐れ入りますが、ご検討のほどよろしくお願いいたします。

ShuntaToda committed:
add directory type

ShuntaToda committed:
add formattedDirData

ShuntaToda committed:
delete useEffect import

ShuntaToda committed:
show element directories

ShuntaToda committed:
add handleDirectoryClick

ShuntaToda committed:
add copy text

ShuntaToda committed:
add copy path

ShuntaToda committed:
delete comment

kondo0602 commented:
@ShuntaToda 
気付くのが遅れて申し訳ありません、確認させていただきます！

kondo0602 commented on src/components/FormattedTextDisplayArea.tsx:
```diff
@@ -1,23 +1,77 @@
+	const { text, setText, formattedText, formattedDirData } = useFormat();
 
+	
+	const createDirectoryElement = (dirData: Directory, dirIndex: number) => {
```
@ShuntaToda 
コンポーネントの中にロジックが出てくるのは避けたいため、このような処理はカスタムフックに切り出すのが良いかと思います！
このプロジェクトでは`/hooks`ディレクトリの中にロジックを切り出しているため、この関数も`generateFormattedLine`のような関数に切り出すことが出来ると思います。

ShuntaToda replied:
generateFormattedLineという関数に切り出して、hook(useDirectories)から呼び出すようにしました。

kondo0602 commented on src/components/FormattedTextDisplayArea.tsx:
```diff
@@ -1,23 +1,77 @@
+	
+	const createDirectoryElement = (dirData: Directory, dirIndex: number) => {
+		
+		const resultPrefixs = []
```
@ShuntaToda 
スペルミスの警告が出ているので修正をお願いします！
```suggestion
		const resultPrefixes = []
```

kondo0602 commented on src/components/FormattedTextDisplayArea.tsx:
```diff
@@ -1,23 +1,77 @@
+		return (
+			<>
+				{resultPrefixs.map((item,index) => <div key={`${dirData.dirName}-${index}`}>{item}</div>)}
+				<div className="cursor-pointer hover:bg-slate-200 rounded" onClick={() => handleDirectoryClick(dirData,dirIndex)}>{dirData.dirName}</div>
+			</>
```
@ShuntaToda 
一つのコンポーネントの中にまた別のコンポーネントが出てくるのは避けたいです！

ここでいうとFormattedTextDisplayAreaの各行を今回作っていただいたこのdiv要素（clickableLineのような感じでコンポーネントとして切り出すのが良さそうです）に置き換え、さらにそのclickableLineの中からロジック部分をgenerateFormattedLineのようなカスタムフックとして呼び出すと綺麗に分けられそうな気がします。

kondo0602 replied:
@ShuntaToda 
またdiv要素にonClickを渡していることでアクセシビリティ上の警告が出ています。
キーボード操作に対応させる or クリック可能な要素に変更する対応が必要そうです。

ShuntaToda replied:
clickableLineコンポーネントを作成しました。
その中でgenerateDirectoryLineDataを呼び出し、表示に必要なデータを取得しています。

ShuntaToda replied:
キーボード操作の部分に関してはbuttonタグを使うことで解決しました。

kondo0602 commented on src/components/FormattedTextDisplayArea.tsx:
```diff
@@ -1,23 +1,77 @@
+		)
+	}
+
+	const getFullPath = (clickedItem: Directory, clickedIndex: number): string => {
```
@ShuntaToda 
こちらも同様に`/hooks`ディレクトリに切り出せそうです。

またこういった関数については簡単なテストがあるとロジックの正当性を確認しやすくて助かります！
（`/hooks`ディレクトリの中の関数には簡単なテストを書いてあるのでよければ参考にしてみてください）

ShuntaToda replied:
こちらの関数なのですが、あえてDirectoryのオブジェクトを受け取って計算する必要がないかと改めて思いました。
なので、parseTextToDirectories関数の中でDirectory[]を作る際にプロパティの一つとしてpathを設定するようにしました。

テストに関してはあまり経験がなかったのですが、とりあえずできる限り書いてみましたので、ご確認いただけると幸いです。

kondo0602 commented on src/components/hooks/formatTextAsDirectoryTree/formatTextAsDirectoryTree.ts:
```diff
@@ -45,8 +47,21 @@ export const generatePrefix = (
 
-export const formatTextAsDirectoryTree = (text: string): string => {
-	return separateLines(text)
+interface formatTextAsDirectoryTree {
+	formattedDirData: Directory[],
+	formattedText: string
+}
```
@ShuntaToda 
細かいのですがinterfaceとtypeの混在が気になりました！
特に意図がなければtypeに統一したいです。

kondo0602 commented on src/components/hooks/useFormat.ts:
```diff
@@ -16,9 +16,10 @@ const defaultValue = `/
+  
 	const debouncedValue = useDebounce(text, { wait: 300 });
-	const formattedText = formatTextAsDirectoryTree(debouncedValue);
+	const {formattedDirData,formattedText} = formatTextAsDirectoryTree(debouncedValue);
```
@ShuntaToda 
フォーマットがかかっていないように見えます！
静的解析にはbiomeを使用しているため`pnpm check`を実行してコードのフォーマットと指摘事項の対応をお願いします！

kondo0602 reviewed (commented):
@ShuntaToda 
とても面白い機能の提案ありがとうございます！
レビューしましたので、指摘事項の確認をお願いします 🙇🏼 

便利そうなので是非マージしたいです。
対応が難しそうであれば時間がある時に自分の方で引き継いで修正しようと思いますので、その旨を伝えていただけると助かります！

kondo0602 reviewed (commented):
@ShuntaToda 
とても面白い機能の提案ありがとうございます！
レビューしましたので、指摘事項の確認をお願いします 🙇🏼 

便利そうなので是非マージしたいです。
対応が難しそうであれば時間がある時に自分の方で引き継いで修正しようと思いますので、その旨を伝えていただけると助かります！

ShuntaToda committed:
interfaceをtypeに変更

ShuntaToda committed:
prefixesのスペルを修正

ShuntaToda committed:
useDirectoriesを作成

add useDirectories

move getFullPath

rename index

ShuntaToda committed:
useDirectoriesからロジックを分割

add parseTextToDirectories

update useDirectories

ShuntaToda committed:
クリック処理の実装をhooksに分けて要素をbuttonにする

delete directories

add generateDirectoryLineData

update ClickableLIne

delete import Directory

delete createDirectoryElement

fix handleClick

fix

move handleDirectoryClick

delete directory import

add type

npm run check

fix key

ShuntaToda committed:
Directoryにpathを追加、parseTextToDirectories内でpathを設定

add path property

delete getFullPath

ShuntaToda committed:
generateDirectoriesにuseCallbackを追加

ShuntaToda committed:
npm run check

ShuntaToda committed:
ロジックのテストを追加

add parseTextToDirectories.test

add generateDirectoryLineData test

ShuntaToda committed:
npm run checkを実行

ShuntaToda commented:
@kondo0602 
すごく丁寧なレビューをして頂きとても感謝しています。僕自身の成長もかなりできました。
最初のプルリク時点で対応できるはずの修正点がいくつもあり、申し訳なかったと感じています。
また、修正点のヒントが豊富にあったことで修正自体もやりやすかったです。
経験が浅く、今回の修正でどこまで希望通りにできているかは想定できない部分がありますが、ご確認いただけたら幸いです。

kondo0602 commented:
@ShuntaToda
早速の対応ありがとうございます！
時間のあるときに確認させていただきます。

kondo0602 commented:
@ShuntaToda 
少しだけレビューがしづらく、お手数ですがコミットの整理をお願いしたいです！

以下2点が理由です。

## 自分が行った指摘に対して、どのコミットでどのように直したかわからないため

現在、ご自身の手元で行った作業のコミットログをそのままPRとしていただいていると思います。

初回の指摘結果を取り込むために30コミット程していただいていますが、自分目線だと自分が行った指摘に対してどこがどのように修正されたかがわからないため、もう一度コード全体を読む必要があります。

自分の場合は 👇🏼  こんな感じでいただいた指摘に対して1コミットで直して、もらった指摘をどう取り込んだかがレビュアにわかりやすいようにしています。

<img width="524" alt="スクリーンショット 2024-08-24 22 19 46" src="https://github.com/user-attachments/assets/ddedb450-91e6-4f3d-96ac-a39020a6a85d">

## 後からコミットログを見返したときに不要なコミットが多く含まれておりノイズになるため

例えばこの[コミットログ](https://github.com/kondo0602/dir-maker/pull/2/commits/9a05e480daaea14d6e946685115276e413b9099a)はプロダクションコードに残る必要があるものでしょうか？

おそらく実際にこのDirectoryを使用しなくなったタイミングがあって、そのときに消し漏れたものを別コミットで消したものではないかと思います。

手元で作業されている時のコミットログは作業しやすいように進めるのが良いと思いますが、最終的にプロダクションコードに含まれるコミットは意味のある単位で・意味のあるコミットコメントが残っていると嬉しいです！

（自分も当時人に見せるものではないと思って実装しており、過去のコミットログが汚いため説得力がないですが...）

以上、確認よろしくお願いします。

ShuntaToda commented:
@kondo0602 
再度丁寧な解説をして頂きありがとうございます🙇
返信が遅くなり申し訳ありません。

調整していてレビュアーの視点を考慮することの重要性を学びました。
コミットをレビューしやすいよう調整しましたが、コーディングの影響で指摘に対して1コミットにまとめることができませんでした。

ご多忙とは思いますが、ご確認いただけると幸いです。

kondo0602 commented:
@ShuntaToda 
ありがとうございます、週末に確認させていただきます！

kondo0602 commented:
@ShuntaToda 
対応ありがとうございます、大変見やすくなったと思います 👍🏼 

useEffectの使用やUIで少しだけ気になるところがあるため、一度developブランチにマージ => 自分の方で少しだけ手直ししてmainブランチにマージとさせていただきます！
developブランチからmainブランチに混ぜる時はPRを作成するので、その折にTodaさんにもお知らせさせていただきます。

ShuntaToda commented:
@kondo0602 
ご確認ありがとうございます。
修正していただけるとのこと承知しました！
こういったパブリックリポジトリに機能追加をするのは初めてでしたが、kondoさんの丁寧な対応のおかげでなんとか完了させることができました。
自分自身の成長にもつながり感謝しかありません！ありがとうございました😄
